File: backend/api/expense.py
from datetime import datetime
from typing import List

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from database import get_db
from models import User
from schemas import ExpenseBatch, ExpenseResponse
from services import expense_service
from api.dependencies import get_current_user


router = APIRouter(tags=["expense"])


@router.post("/expense")
def create_expense(
    data: ExpenseBatch,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    results = [expense_service.create_expense(db, expense, str(current_user.id)) for expense in data.expenses]
    return {"success": True, "count": len(results), "data": results}


@router.get("/expense", response_model=List[ExpenseResponse])
def list_expense(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return expense_service.get_expenses(db, str(current_user.id))


@router.delete("/expense/{expense_id}")
def delete_expense(
    expense_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    deleted = expense_service.delete_expense(db, expense_id, str(current_user.id))
    return {"success": deleted}


@router.get("/summary")
def summary(db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return expense_service.get_summary(db, str(current_user.id))


@router.get("/expense/month")
def month_summary(
    year: int,
    month: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if month < 1 or month > 12:
        return {"success": False, "message": "month must be between 1 and 12"}
    return expense_service.get_month_summary(db, str(current_user.id), year, month)


@router.get("/expense/query")
def query_expense(
    start_time: datetime,
    end_time: datetime,
    category: str | None = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return expense_service.get_query_summary(db, str(current_user.id), start_time, end_time, category)

File: frontend/src/api/expense.ts
import { request } from './client';
import type { ExpenseItem, SummaryResponse, QuerySummaryResponse } from '../types/expense';

/** GET /expense —— 当前用户全部账单（后端不做排序，前端自行按时间排） */
export function listExpenses(): Promise<ExpenseItem[]> {
  return request<ExpenseItem[]>('/expense');
}

/** DELETE /expense/{id} —— 不存在/越权时后端返回 200 + {success:false} */
export function deleteExpense(id: number): Promise<{ success: boolean }> {
  return request<{ success: boolean }>(`/expense/${id}`, { method: 'DELETE' });
}

/** GET /summary —— 全时段汇总 */
export function getSummary(): Promise<SummaryResponse> {
  return request<SummaryResponse>('/summary');
}

/** GET /expense/month?year&month —— 单月汇总（month 必须 1-12） */
export function getMonthSummary(year: number, month: number): Promise<SummaryResponse> {
  const params = new URLSearchParams({ year: String(year), month: String(month) });
  return request<SummaryResponse>(`/expense/month?${params.toString()}`);
}

/**
 * GET /expense/query?start_time&end_time —— 时间区间汇总。
 * 时间用不带时区的本地 ISO 串（如 2026-09-08T00:00:00），与后端 naive datetime 对齐。
 */
export function getRangeSummary(startTime: string, endTime: string): Promise<QuerySummaryResponse> {
  const params = new URLSearchParams({ start_time: startTime, end_time: endTime });
  return request<QuerySummaryResponse>(`/expense/query?${params.toString()}`);
}
